### PR TITLE
Gs improvements

### PR DESCRIPTION
**Pull request by andrewchambers**
With this patch set I have been able to avoid #84 for multiple hours now.

### CODE REVIEW

**Comment by andrewchambers** [comment]
@Nikratio One question I have, the access_token dict is not locked, I guess it relies on the GIL for safe reads, is this intentional?

File: src/s3ql/backends/gs.py [hunk 1]
```diff
@@ -451,29 +488,14 @@ def write_fh(self, fh, key: str, md5: bytes,
...
-        try:
-            self.conn.write(body_prefix)
-            while True:
-                buf = fh.read(BUFSIZE)
-                if not buf:
-                    break
-                self.conn.write(buf)
-                md5_run.update(buf)
-            self.conn.write(body_suffix)
-        except ConnectionClosed:
```
Comment (andrewchambers) [inline src/s3ql/backends/gs.py]: @Nikratio This logic seemed to be copy pasted and not make much sense, so I just removed it.

**Comment by andrewchambers** [comment]
I will squash history and get rid of merge commits after review.

File: src/s3ql/backends/gs.py [hunk 2]
```diff
@@ -130,23 +130,54 @@ def  __call__(self, url: str, method: str = 'GET', body=None,
...
+            return Namespace(status=resp.status, headers=resp.headers,
+                         data=conn.readall())
         except (dugong.ConnectionClosed, dugong.InvalidResponse, dugong.UnsupportedResponse,
                 dugong.ConnectionTimedOut, dugong.HostnameNotResolvable,
                 dugong.DNSUnavailable, ssl.SSLError) as exc:
             raise g_auth.exceptions.TransportError(exc)
+        finally:
+            conn.disconnect()
 
-        return Namespace(status=resp.status, headers=resp.headers,
```
Comment (andrewchambers) [inline src/s3ql/backends/gs.py]: - [ ] revert pointless move of return

File: src/s3ql/backends/gs.py [hunk 3]
```diff
@@ -225,6 +256,7 @@ def __init__(self, options):
             raise
         self._parse_json_response(resp)
 
+
```
Comment (andrewchambers) [inline src/s3ql/backends/gs.py]: - [ ] revert pointless whitespace insertion.

**Comment by Nikratio** [comment]
Note to self: there's also the problem that google.auth sometimes wants to use the timeout argument. I think this needs to be fixed as well.

**Comment by andrewchambers** [comment]
Another approach is to use the adc.valid property which checks timestamps, it is a bit annoying to distinguish between a real 401 and a time related one when multiple threads may be touching the ADC. I also am unsure about non ADC oauth expiration.

**Comment by Nikratio** [comment]
Quoting from issue #84:

> -     The auth requestor pool is not cleaning up disconnected connections.

This should be fixed with commit f76a8.

> -     Google auth transport errors are not marked as temporary failures.

Right.

> -     Expired tokens are not being removed from the access_token map.

Could you elaborate? I don't see where this should be happening.  _get_access_token() should overwrite the expired token, so why the need for explicit removal?

> -     401 errors seem to be happening in write_fh where it is not checked for retry.

`write_fh` has a `@retry` decorator - could you elaborate?

File: src/s3ql/backends/gs.py (hunk 1 above)
Comment (Nikratio) [reply src/s3ql/backends/gs.py]: Without this logic, we will get a generic dugong exception that tells us nothing beyond "writing to the server failed". This code tries to read a server response, which allows us to raise an exception with more details provided by the server.

Comment (andrewchambers) [reply src/s3ql/backends/gs.py]: I can't find where in the code the response is used? at the call site it is ignored.

File: src/s3ql/backends/gs.py [hunk 4]
```diff
@@ -452,29 +487,14 @@ def write_fh(self, fh, key: str, md5: bytes,
...
-        except ConnectionClosed:
-            # Server closed connection while we were writing body data -
-            # but we may still be able to read an error response
-            try:
-                resp = self.conn.read_response()
-            except ConnectionClosed: # No server response available
-                pass
-            else:
-                if resp.status >= 400: # Got error response
-                    return resp
```
Comment (andrewchambers) [inline src/s3ql/backends/gs.py]: @Nikratio (I think I accidentally deleted your comment and can't get it back.) The resp returned here is never used at the call site, that is why it didn't make sense to me.

Comment (Nikratio) [reply src/s3ql/backends/gs.py]: Good catch. The `return` here is completely bogus, presumably some leftover debugging code. Committed a fix.

**Pull request closed without merging**